examples/transformerconfigs/README.md

```diff
@@ -119,44 +119,472 @@ commonAnnotations:
 
 ## Name reference transformer
 
-Name reference transformer's configuration is different from all other transformers. It contains a list of `nameReferences`, which represent all of the possible fields that a type could be used as a reference in other types of resources. A `nameReference` contains a type such as ConfigMap as well as a list of `fieldSpecs` where ConfigMap is referenced in other resources. Here is an example:
+`nameReference` Transformer is used to tie a target resource's name to a list of other resources' referrers' names.
+Once tied, the referrers' names will change alongside the target name via transformers like `namePrefix` and `nameSuffix`
 
+### Usage
+  - The syntax `nameReference` should be written in the `configurations` files, not directly in `kustomization.yaml`
+  - kustomize has a set of builtin nameReference, and you don't need to write additional configs to use those nameReference. Check the full list [here](#buildin-namereference).
+  - The referrer's `name` has to match the target's `name`. Otherwise, the referrer's name won't be changed.
+  - `nameReference` should be used together with other name-changing transformers. Using it alone won't make any changes.
+
+### Example
+
+kustomization.yaml
 ```yaml
-kind: ConfigMap
-version: v1
-fieldSpecs:
-- kind: Pod
-  version: v1
-  path: spec/volumes/configMap/name
-- kind: Deployment
-  path: spec/template/spec/volumes/configMap/name
-- kind: Job
-  path: spec/template/spec/volumes/configMap/name
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+
+resources:
+  # resources.yaml contains a Gorilla and a AnimalPark
+  - resources.yaml
+
+# Add name prefix "sample-" to the Gorilla's and AnimalPark's name
+namePrefix: sample-
+
+configurations:
+  # Tie target Gorilla to AnimalPark's referrer spec.gorillaRef.
+  # Once Gorilla name is changed, the AnimalPark referrerd Gorilla name will be changed as well.
+  - nameReference.yaml
+```
+nameReference.yaml
+```yaml
+nameReference:
+  - kind: Gorilla
+    fieldSpecs:
+      - kind: AnimalPark
+        path: spec/gorillaRef/name
+```
+
+resources.yaml
+```yaml
+apiVersion: animal/v1
+kind: Gorilla
+metadata:
+  name: gg
+---
+apiVersion: animal/v1
+kind: AnimalPark
+metadata:
+  name: ap
+spec:
+  gorillaRef:
+    name: gg
+    kind: Gorilla
+    apiVersion: animal/v1
+```
+Output of `kustomize build`
+```yaml
+apiVersion: animal/v1
+kind: AnimalPark
+metadata:
+  name: sample-ap # changed by `namePrefix`
+spec:
+  gorillaRef:
+    apiVersion: animal/v1
+    kind: Gorilla
+    name: sample-gg # changed by `nameReference`
+---
+apiVersion: animal/v1
+kind: Gorilla
+metadata:
+  name: sample-gg # changed by `namePrefix`
 ```
 
-Name reference transformer's configuration contains a list of `nameReferences` for resources such as ConfigMap, Secret, Service, Role, and ServiceAccount. Here is an example configuration:
+### buildin NameReference
 
 ```yaml
 nameReference:
```

Comment:
Is this a copy of https://github.com/kubernetes-sigs/kustomize/blob/master/api/konfig/builtinpluginconsts/namereference.go? 

Would it make more sense to just link to https://github.com/kubernetes-sigs/kustomize/blob/master/api/konfig/builtinpluginconsts/namereference.go instead of copying here? In the case that `namereference.go` changes, we will now have to update it in two places and these docs may quickly become stale

Reply:
It is copied from that file. 
Yeah, I was originally thinking the same, like referring to the source code. Two reasons make me decide to copy the code here:
1.  Exposing the builtin nameReferences to README makes this feature more searchable. It is very rare to see examples to nameReference, i guess one reason might be the invisibility of usage and incomplete docs.   
2. konfig will be refactored to internal for V1, so referring the HEAD link will be deprecated soon and referring a versioned code like https://github.com/kubernetes-sigs/kustomize/blob/ed3200e4f515c43259f0c1bcc5811a869f3551f4/api/konfig/builtinpluginconsts/namereference.go will be the same as "need to update in two places". A third option is to add a "./const" directory and keep the file path unchanged. So even with v1 or other refactoring changes, it is always safe to refer the HEAD of `./const`.   

I'm totally happy to refer to master/api/konfig/builtinpluginconsts/namereference.go as well. What do you think? @natasha41575 @KnVerey

Reply:
I see, thanks for explaining. I think it is fine if we keep a copy in the README then. Can we add a comment to the top of https://github.com/kubernetes-sigs/kustomize/blob/master/api/konfig/builtinpluginconsts/namereference.go that says something like, `there is a copy of these fieldspecs in https://github.com/kubernetes-sigs/kustomize/blob/master/examples/transformerconfigs/README.md`, so that whenever we update `namereference.go`, we can have a reminder to also update these docs?